### DISCUSSION TITLE
Generate Release notes with work items not associated to any Pull Requests.

### DISCUSSION THREAD

**Discussion by Imran109** [rfennell/AzurePipelines#1117] [Q&A] [answered]
Hello, is there a way to get access to the work items of a project that belongs to a specific area and iteration path that are not necessarily associated to any Pull Requests. I tried looking out for this is samples and documentation but couldn't find any help. 

What we tried (Since it says global list of work items, I assumed it would fetch every items in a project, but its just fetching the linked work items to the repo. And our use case is to get all the work items of a project on some filter(By type, area and iteration) in the same project)

```
# Global list of WI ({{workItems.length}})
              {{#forEach workItems}}
              {{#if isFirst}}## Associated Work Items (only shown if  WI) {{/if}}
              *  **{{this.id}}**  {{lookup this.fields 'System.Title'}}
                - **WIT** {{lookup this.fields 'System.WorkItemType'}} 
                - **Tags** {{lookup this.fields 'System.Tags'}}
              {{/forEach}}
```

**Comment by rfennell** [comment]
You have two options I think

## Using If blocks
 You would need to add `{{#if ` blocks  something like (I have not tested this but it should be close)
 
```
 # Global list of WI ({{workItems.length}})
{{#forEach workItems}}
{{#if (and (contains (lookup this.fields 'System.AreaPath') 'A path') (contains (lookup this.fields 'System.IterationPath') 'A IterationPath'))}}
    {{#if isFirst}}## Associated Work Items (only shown if  WI) {{/if}}
        *  **{{this.id}}**  {{lookup this.fields 'System.Title'}}
         - **WIT** {{lookup this.fields 'System.WorkItemType'}} 
         - **Tags** {{lookup this.fields 'System.Tags'}}
{{/if}}
{{/forEach}}
```

## Write a custom extension
You could write some Javascript to write and enhanced filtering foreach. [This sample](https://github.com/rfennell/AzurePipelines/blob/main/SampleTemplates/XplatGenerateReleaseNotes%20(Node%20based)/Version%203/each_with_sort_by_field.js) does a filter loop, but would need some editing

**Reply by Imran109** [reply]
@rfennell Thanks for the response, as mentioned looks like this would only fetch the WIs linked to that repository or any PRs in that specific repository. We need to fetch all the items that are also not associated to any pull requests or repositories. Basically we need to access all the WIs of a project and make a filter on them.

**Reply by rfennell** [reply]
There is no current support for including WI that are not either

- Associated with a commit/changeset
- A PR
- A Builds (manually and automatically)
- Is a parent of one of the above

There is feature request #1037 that would maybe address your need, but no immediate plan to add that feature. Always happy to consider PRs

**Reply by rfennell** [reply]
Change of plan, I am activally working on #1037 should be done soon

**Comment by rfennell** [comment, marked as answer]
This feature is available in 3.71.x

There is a new parameter where you can provide the WHERE part of the WIQL. This should be edited to select the WI you require.

```
 - task: richardfennellBM.BM-VSTS-XplatGenerateReleaseNotes-DEV1.XplatGenerate-Release-Notes.XplatGenerateReleaseNotes@3
    inputs:
      wiqlWhereClause: '[System.TeamProject] = "GitHub" and [System.WorkItemType] = "Product Backlog Item"'
```
Note you cannot use `@project`, `@currentiteration` or `@me`, but `@today` is ok. To work out the WHERE clause I recommend using the [WIQL Editor](https://marketplace.visualstudio.com/items?itemName=ottostreifel.wiql-editor) extension

The results of this WIQL will be available in a new independent template array `queryWorkItems` i.e. it is completely separate from all the other WI arrays.

```
# WIQL list of WI ({{queryWorkItems.length}})
{{#forEach queryWorkItems}}
   *  **{{this.id}}** {{lookup this.fields 'System.Title'}}
{{/forEach}}
```

**Reply by Imran109** [reply]
Thanks a lot for the quick response 😉. We are now using this feature in our project.

**Reply by Imran109** [reply]
Hey @rfennell , Any plans to include **@currentIteration** to work in where clause

**Reply by rfennell** [reply]
Sorry no,  the `@currentiteration` macro is not supported by the REST API WQL call, if you try you get the following error because the API does not know what team and hence `areapath` that should be used

> {"$id":"1","innerException":null,"message":"TF401347: Invalid tree name given for work item -1, field 'System.IterationPath'.","typeName":"Microsoft.TeamFoundation.WorkItemTracking.Server.WorkItemFieldInvalidTreeNameException, Microsoft.TeamFoundation.WorkItemTracking.Server, Version=14.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a","typeKey":"WorkItemFieldInvalidTreeNameException","errorCode":600171,"eventId":3200}
